### PR TITLE
ENH: Implement multi-column `DataFrame.quantiles`

### LINKED ISSUE #43881: ENH: Add support for multi-column quantiles of DataFrame
#### Is your feature request related to a problem?

For dataframes, Pandas currently only supports per-column quantiles; that is, given `df[['c', 'a']].quantile(...)`, Pandas will compute the individual quantiles for columns `c` and `a`: 

```python
>>> df = pd.DataFrame({'a': [1, 0, 11, 12, 2], 'b': [1, 2, 3, 4, 5], 'c': [0, 1, 5, 2, 3]})
>>> df[['c', 'a']].quantile([0, 0.5, 1])
       c     a
0.0  0.0   0.0
0.5  2.0   2.0
1.0  5.0  12.0
```

It would be nice if Pandas also supported multi-column quantiles; that is, given  `df[['c', 'a']].quantiles(...)`, Pandas would compute the quantiles for the dataframe sorted by _all_ columns. This is currently implemented by cuDF's dataframe:

```python
In [11]: gdf = cudf.DataFrame({"a": [1, 1, 1, 1, 1], "b": [5, 4, 3, 2, 1]})
In [12]: gdf[["a", "b"]].quantiles([0, 0.5, 1])
Out[12]: 
     a  b
0.0  1  1
0.5  1  3
1.0  1  5
```

#### Describe the solution you'd like

I imagine the addition of multi-column quantiles support could happen in two ways:

1. the addition of a default kwarg to `DataFrame.quantile` to specify whether or not we want multi-column quantiles
2. the addition of a new method to compute multi-column quantiles independent from the logic of `quantile` 

In either case, my preference here would be to have this functionality accessible via `DataFrame.quantiles`, to maintain consistency with cuDF.

#### API breaking implications

I can't think of any breakages this would cause, as long as any direct changes to `quantile` ensure that the original behavior is maintained by default.

#### Describe alternatives you've considered

This could be accomplished by sorting the dataframe by all columns and then indexing based on manually computed quantiles, but I imagine there's a more performant way to do this.

#### Additional context

If this functionality were added, along with a multi-columnar `searchsorted`, it would enable Dask dataframes to compute `sort_values` with multiple sort-by columns, using an algorithm roughly similar to [that of dask-cudf](https://github.com/rapidsai/cudf/blob/ffb3814089c9a0c3c332d9a36714ab853a48ca9d/python/dask_cudf/dask_cudf/sorting.py#L208-L254).

### PR DESCRIPTION
**Pull request by charlesbluca**
Rough attempt at implementing cuDF's `DataFrame.quantiles`; shares a lot of common logic with `sort_values`, as the indexer that sorts the dataframe by all columns is ultimately what is used to grab the desired quantiles.

cc @quasiben @rjzamora

- [ ] closes #43881
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [ ] whatsnew entry

### CODE REVIEW

**Comment by quasiben** [comment]
@TomAugspurger this is an attempt at trying to get multi-column sorting working in Dask which requires a multi-column quantile

File: pandas/core/frame.py
Comment (rjzamora) [inline pandas/core/frame.py:10467]: [cudf's `DataFrame.quantiles`](https://github.com/rapidsai/cudf/blob/7e112689eb33b04d81b1b70ea7c7ba5df08cdfe0/python/cudf/cudf/core/dataframe.py#L5265) doesn't support a `numeric_only` argument, so the effective default is `numeric_only=False`. Any chance we could modify the default here? Is this meant to align with `quantile` arguments?

Note that I can understand the argument for `numeric_only=True`, but it may add a bit of extra pain in Dask :)

Comment (charlesbluca) [reply pandas/core/frame.py:10467]: Yeah, I did this to align with the default for `quantile` - happy to change to `False` if it make sense to the devs

**Comment by TomAugspurger** [comment]
I'm not sure about the name... I worry about having both a `.quantile` and `quantiles`, and I don't think that quantile**s** really conveys what's different about this method. To me the plural version sounds like you're able to compute multiple quantiles at once, which is already supported by passing a list to `quantile`.

So if we do this, I'd suggest a name like `multi_column_quantile` or `multi_quantile` or `quantile_table` (the last one mirrors a proposed API for `groupby.agg` on a table). Or perhaps we make this a keyword for `quantile`, like `table=True` or `columnar=False`? Because IIUC the output type (Series or Frame) and labels will be identical `.quantile`, the only difference is the values?

**Comment by quasiben** [comment]
@shwina, do you think we could alias `quantiles` in cuDF if we do end up changing the name here ?

**Comment by shwina** [comment]
Yes -- that should be OK for cuDF. I also like `multi_quantile` over `quantiles`

**Comment by charlesbluca** [comment]
Cool! In that case I'm going to rename this method `multi_quantile`, and we can follow up with a rename / aliasing on cuDF?

**Comment by jreback** [comment]
@charlesbluca this hasn't received any scruity yet. -1 on adding methods directly like this.

**Review by jreback — CHANGES_REQUESTED** [review]
tests are the first thing that is needed

**Comment by rjzamora** [comment]
> -1 on adding methods directly like this.

@jreback - Is your preference to add a new option to the existing `quantile` method. Or are you saying that down-stream libraries should be implementing logic like this themselves?

**Comment by jreback** [comment]
i would add the argument `method='single|table'` which is what we do for `.rolling` and friends

File: pandas/core/frame.py
Comment (jreback) [inline pandas/core/frame.py:10471]: this should all be in the quantile method in algos not here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py
Comment (charlesbluca) [reply pandas/core/frame.py:10471]: For clarity, do you mean that this code should be in a new `quantile` method in algos that handles the table case, or in `BlockManager.quantile`, where it looks like the internal implementation of `DataFrame.quantile` resides?

**Comment by charlesbluca** [comment]
Currently blocked on handling for sparse arrays - I am using `BlockManager.take` to perform the actual quantile operation, which retains sparse dtypes where `BlockManager.quantile` would not. This causes issues when handling non-list-like `qs` as an `iloc` across columns is required, which does not seem to be possible with sparse dtypes:

```python
    def test_quantile_sparse(self, df, expected):
        # GH#17198
        # GH#24600
>       result = df.quantile(interpolation="nearest", method="table")

pandas/tests/frame/methods/test_quantile.py:591: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pandas/core/frame.py:10475: in quantile
    return res.iloc[0]
pandas/core/indexing.py:957: in __getitem__
    return self._getitem_axis(maybe_callable, axis=axis)
pandas/core/indexing.py:1509: in _getitem_axis
    return self.obj._ixs(key, axis=axis)
pandas/core/frame.py:3483: in _ixs
    new_values = self._mgr.fast_xs(i)
pandas/core/internals/managers.py:974: in fast_xs
    result = cls._empty((n,), dtype=dtype)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'pandas.core.arrays.sparse.array.SparseArray'>, shape = (2,), dtype = Sparse[int64, 0]

    @classmethod
    def _empty(cls, shape: Shape, dtype: ExtensionDtype):
        """
        Create an ExtensionArray with the given shape and dtype.
        """
        obj = cls._from_sequence([], dtype=dtype)
    
        taker = np.broadcast_to(np.intp(-1), shape)
        result = obj.take(taker, allow_fill=True)
        if not isinstance(result, cls) or dtype != result.dtype:
>           raise NotImplementedError(
                f"Default 'empty' implementation is invalid for dtype='{dtype}'"
            )
E           NotImplementedError: Default 'empty' implementation is invalid for dtype='Sparse[int64, 0]'

pandas/core/arrays/base.py:1492: NotImplementedError
```

A few follow up questions here:

- Do we want `quantile(method="table")` to not retain sparse dtypes to match `quantile(method="single")`?
- If not, are there any blockers to `iloc`ing across sparse columns that can be addressed to unblock this PR?

**Comment by charlesbluca** [comment]
Still interested in working on this, currently blocked by the handling for sparse arrays, specifically if want to retain sparse dtypes for `quantile(method="table")`

**Comment by charlesbluca** [comment]
@TomAugspurger if you get a chance could I get a follow-up review on this? Also a little confused by the test failures here - in trying to resolve the single test failures with sparse dtypes in runs like [this](https://github.com/pandas-dev/pandas/runs/4807168239?check_suite_focus=true), I seem to create a [lot more failures](https://github.com/pandas-dev/pandas/runs/4822740902?check_suite_focus=true) that I'm unable to replicate locally.

**Comment by jreback** [comment]
@charlesbluca if you can merge master and we'll have a look

File: pandas/core/frame.py
Comment (mroeschke) [inline pandas/core/frame.py:11151]:
```suggestion
        method: Literal["single", "table"] = "single",
```

File: pandas/core/frame.py
Comment (mroeschke) [inline pandas/core/frame.py:11309]: Is there a test for this exception?

File: pandas/tests/frame/methods/test_quantile.py
Comment (mroeschke) [inline pandas/tests/frame/methods/test_quantile.py:645]: Probably okay to import at the top

File: pandas/tests/frame/methods/test_quantile.py
Comment (mroeschke) [inline pandas/tests/frame/methods/test_quantile.py:662]: For these comments, could you turn each one into dedicated testing functions?

File: pandas/tests/frame/methods/test_quantile.py
Comment (mroeschke) [inline pandas/tests/frame/methods/test_quantile.py:933]: Could you use `pytest.mark.paremeterize` for these cases?

File: pandas/core/frame.py
Comment (mroeschke) [inline pandas/core/frame.py:11183]: Just to confirm, the allowed interpolation methods are based on `np.quantile` interpolation support?

File: pandas/core/frame.py
Comment (mroeschke) [inline pandas/core/frame.py:11297]: Did this need to be converted to a list? `data.columns` should be list-like

File: pandas/tests/frame/methods/test_quantile.py
Comment (mroeschke) [inline pandas/tests/frame/methods/test_quantile.py:1160]: To confirm, these are numpy related versions?

File: pandas/tests/frame/methods/test_quantile.py
Comment (charlesbluca) [inline pandas/tests/frame/methods/test_quantile.py:91]: Is this test asserting that the output columns are correct when `method == "table"`?

File: pandas/tests/frame/methods/test_quantile.py
Comment (charlesbluca) [inline pandas/tests/frame/methods/test_quantile.py:98]: Is this test asserting that the output columns are correct when `method == "table"`?

**Review by charlesbluca — COMMENTED** [review]
Thanks for pushing this along @mroeschke 😄 some small questions around the modified tests:

File: pandas/tests/frame/methods/test_quantile.py
Comment (mroeschke) [reply pandas/tests/frame/methods/test_quantile.py]: The result will be Series objects, but I improved the assertions here to compare the entire Series results if the interpolation is linear (& method = single) or compare Series name + index if interpolation is nearest (& method = table)

File: pandas/tests/frame/methods/test_quantile.py
Comment (mroeschke) [reply pandas/tests/frame/methods/test_quantile.py]: The result will be Series objects, but I improved the assertions here to compare the entire Series results if the interpolation is linear (& method = single) or compare Series name + index if interpolation is nearest (& method = table)

File: pandas/core/frame.py
Comment (jreback) [inline pandas/core/frame.py:11289]: maybe i am not getting something, but why isn't this just
```np.asarray(res_df).ravel()``` and then reed to the existing quantile routine?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py
Comment (mroeschke) [reply pandas/core/frame.py:11289]: 1. I think that approach would not work for DataFrames with mixed dtypes
2. For the limited set of interpolation methods supported (to start) in this PR, I think this approach is more performant as only quantile indices are calculated followed by a `take`.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @charlesbluca really nice! (and @mroeschke for pushing over the line)

**Pull request merged**
